### ISSUE TITLE
Mail server, I can no longer log into Postfixadmin

### ISSUE THREAD

**Issue by sovkutsan** [postfixadmin/postfixadmin#776]
I have configured the mail server and it is working. But after I redirected to https://, I can’t log into Postfixadmin even on the server itself.

Invalid token! (CSRF check failed)

I configured it according to standard instructions, there is a Let's Encrypt certificate.

I have Ubuntu22.04
Apache port 80
Nginx redirects from 8080 (the 80th is occupied by Apache) to 443

Here is my simple Nginx config

(Until I added the line "Redirect / https://mail.example.com/" to the /etc/apache2/sites-available/000-default.conf file"

The redirect configured in nginx did not work, no matter what I did)

I spent 3 days searching, but I don’t understand anything:
About tokens and sessions, so I don’t know how to use even the information that I find.

I haven’t set up a mail server for 10 years, before it wasn’t necessary to do https://, and this error about CSRF has left me helpless, I absolutely don’t understand what to do.

I will be grateful for any help.

**Comment by sovkutsan** [comment]
server {
        listen 8080; 
        server_name mail.example.com;
    return 301 https://$host$request_uri;
    }

server {    
        listen 443 ssl;
        
          
       #location /auth-basic {
       #       auth_basic            "Basic Auth";
       #       auth_basic_user_file  "/etc/nginx/auth.basic";
       # }


          root /var/www/html;
         #root /var/www/postfixadmin/public/;
         
        # Add index.php to the list if you are using PHP
        index index.php index.html index.htm index.nginx-debian.html;

   
                               
        #  listen 443 ssl; 
          ssl_certificate /etc/letsencrypt/live/mail.example.com/fullchain.pem;
          ssl_certificate_key /etc/letsencrypt/live/mail.example.com/privkey.pem;
          include /etc/letsencrypt/options-ssl-nginx.conf;
          ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;
          # return 404; # managed by Certbot
       
         
      
   
        location / {
        
    
                # First attempt to serve request as file, then
                # as directory, then fall back to displaying a 404.
                 try_files $uri $uri/ =404;   
                
                             
                # proxy_pass http://localhost:8080;
                # proxy_http_version 1.1;
                # proxy_set_header Upgrade $http_upgrade;
                # proxy_set_header Connection 'upgrade';
                # proxy_set_header Host $host;
                # proxy_cache_bypass $http_upgrade;
        }

        location ~ \.php$ {
     
            #auth_basic            "Protection";
            # auth_basic_user_file  "/etc/nginx/auth.login";
        
            set $root_path /var/www/html;
            fastcgi_pass unix:/run/php/php8.1-fpm.sock;
            fastcgi_index index.php;
            fastcgi_param SCRIPT_FILENAME $root_path$fastcgi_script_name;
            include fastcgi_params;
            fastcgi_param DOCUMENT_ROOT $root_path;
            
        }
        
  
}

**Comment by sovkutsan** [comment]
/etc/apache2/sites-available/000-default

<VirtualHost *:80>
	# The ServerName directive sets the request scheme, hostname and port that
	# the server uses to identify itself. This is used when creating
	# redirection URLs. In the context of virtual hosts, the ServerName
	# specifies what hostname must appear in the request's Host: header to
	# match this virtual host. For the default virtual host (this file) this
	# value is not decisive as it is used as a last resort host regardless.
	# However, you must set it for any further virtual host explicitly.
	#ServerName www.example.com

	ServerAdmin webmaster@localhost
	DocumentRoot /var/www/html

	# Available loglevels: trace8, ..., trace1, debug, info, notice, warn,
	# error, crit, alert, emerg.
	# It is also possible to configure the loglevel for particular
	# modules, e.g.
	#LogLevel info ssl:warn

	ErrorLog ${APACHE_LOG_DIR}/error.log
	CustomLog ${APACHE_LOG_DIR}/access.log combined

	# For most configuration files from conf-available/, which are
	# enabled or disabled at a global level, it is possible to
	# include a line for only one particular virtual host. For example the
	# following line enables the CGI configuration for this host only
	# after it has been globally disabled with "a2disconf".
	#Include conf-available/serve-cgi-bin.conf
	

Redirect / https://mail.example.com/

	
	
</VirtualHost>

# vim: syntax=apache ts=4 sw=4 sts=4 sr noet

**Comment by sovkutsan** [comment]
Previously, I added mailboxes on the server through a browser, wrote, (my ip)/postfixadmin/local/login.php)

Now automatically I see https://mail.example.com/postfixadmin/local/login.php)

The Postfixadmin window opens, as well as Roundcube, but it’s impossible to get inside because of (CSRF check failed)

Not a single email setup instruction even mentioned that this check would occur.

**Comment by DavidGoodwin** [comment]
The CSRF check involves a PHP session variable (and cookie, used by PHP's session handler). and the same value is injected into various forms. If they match, then everything's alright.

Reasons for it failing ...

 * Problem with your browser accepting cookies 
 * Problem with PHP sessions or session storage (e.g. a new session is created on each page load or something)
 * Trying to access PostfixAdmin through multiple tabs/windows (I suspect this would trigger the problem).

**Comment by sovkutsan** [comment]
Dear DavidGoodwin
Thanks for your reply.

I didn’t quite understand the phrase (the same meaning is entered in different forms)
Because I always open one form (Postfixadmin or Roundcube)

I always use them separately. Or the Postfixadmin tab or the Roundcube tab.

What does it mean - the same meaning?
Login and Password are different values. Login is always the same (root@example.com)

For Roundcube login (Nik@example.com)

My passwords are the same for both programs.


Thanks for describing the possible reasons for the failure.

1) I have disabled Cookie reception in Firefox.

2) You said - “A new session is created every time the page is loaded,” how could it be different?
Can you give me any advice? - Where is the session storage located?
I want to clean it.

3) I always access PostfixAdmin through one tab.

Now I tried this script: 

<?php
session_start();
if(false == isset($_SESSION['test'])) {
     var_dump("init ... refresh please");
    $_SESSION['test'] = uniqid();
}

var_dump('Session contains', $_SESSION);
?>

Result of running the script:(Launched from /var/www/html/skript.php)

string(23) "init ... refresh please" string(16) "Session contains" array(1) { ["test"]=> string(13) "6541a42b52d61" } 

If I run the script from /var/www/html/postfixadmin/public/skript.php
The result is different:
string(16) "Session contains" array(1) { ["test"]=> string(13) "6541af6216b78" } 

It is not clear in which file these lines are (23,16,13)
Maybe you know?
I found the script on the Internet. It is designed to test php

Session is a login to Postfixadmin? and somewhere else? For example in Roundcube...
To GMail...
Or to any site?
Or is running a script also a session?

I apologize for such stupid questions, I just haven’t come across these concepts before...

What if I disable Apache|
so that he does not touch port 80
And I use this port for Nginx
I thought that two web servers are not necessarily needed together.
I'll also try to disable error output now...

Perhaps the problem is that I have two php
1) Many different php modules for Apache
2) php-fpm8.1 For Nginx

Perhaps the problem is in php-fpm...

Thank you for your attention, I apologize for writing so much. I will be fired if I don’t log into Postfixadmin this week)) The bosses are asking me to create mailboxes already))

Now I'll make all the changes and see what happens.

**Comment by DavidGoodwin** [comment]
Postfixadmin uses a PHP session to store login state and for CSRF stuff.

PHP's sessions REQUIRE a cookie to be sent by the browser. If that cookie changes or is absent (like your browser is blocking cookies) then a new session will be started on each page load, which stops Postfixadmin from being able to store state and therefore it won't work.

On the backend, PHP is linking a value in a cookie (e.g. PHPSESSID=12345abc) with a file on disk (e.g /var/lib/php/sessions/12345abc).

Postfixadmin depends on PHP's session support.

You can't disable cookies in your browser and expect postfixadmin to work.

**Comment by sovkutsan** [comment]
Hello
Today I changed some things, but the understanding has not yet appeared.
in file     etc/php/8.1/fpm/php.ini

I did: display_errors = On

After this, when logging into Postfixadmin, error information appeared, first like this:

Warning: Unknown: POST Content-Length of 131 bytes exceeds the limit of 50 bytes in Unknown on line 0

Warning: session_cache_limiter(): Session cache limiter cannot be changed after headers have already been sent in /var/www/html/postfixadmin/common.php on line 23

Warning: session_name(): Session name cannot be changed after headers have already been sent in /var/www/html/postfixadmin/common.php on line 24

Warning: session_start(): Session cannot be started after headers have already been sent in /var/www/html/postfixadmin/common.php on line 25
Invalid token (session timeout; refresh the page and try again?)

Reloading the page didn't change anything

Then, following your advice, I enabled cookies and installed

error_reporting = E_ERROR


After that, immediately when the window appears (I didn’t even have to enter my login and password)
Now above the postfix admin login window all the time there are these errors:

Deprecated: htmlentities(): Passing null to parameter #1 ($string) of type string is deprecated in /var/www/html/postfixadmin/model/PFASmarty.php on line 137
Deprecated: htmlentities(): Passing null to parameter #1 ($string) of type string is deprecated in /var/www/html/postfixadmin/model/PFASmarty.php on line 137
Deprecated: htmlentities(): Passing null to parameter #1 ($string) of type string is deprecated in /var/www/html/postfixadmin/model/PFASmarty.php on line 137
Deprecated: htmlentities(): Passing null to parameter #1 ($string) of type string is deprecated in /var/www/html/postfixadmin/model/PFASmarty.php on line 137

If you enter your username and password in Postfixadmin, the window disappears and all errors appear
Both the original and the second

Warning: Unknown: POST Content-Length of 131 bytes exceeds the limit of 50 bytes in Unknown on line 0

Warning: session_cache_limiter(): Session cache limiter cannot be changed after headers have already been sent in /var/www/html/postfixadmin/common.php on line 23
Warning: session_name(): Session name cannot be changed after headers have already been sent in /var/www/html/postfixadmin/common.php on line 24
Warning: session_start(): Session cannot be started after headers have already been sent in /var/www/html/postfixadmin/common.php on line 25
Deprecated: htmlentities(): Passing null to parameter #1 ($string) of type string is deprecated in /var/www/html/postfixadmin/model/PFASmarty.php on line 137

Deprecated: htmlentities(): Passing null to parameter #1 ($string) of type string is deprecated in /var/www/html/postfixadmin/model/PFASmarty.php on line 137
Deprecated: htmlentities(): Passing null to parameter #1 ($string) of type string is deprecated in /var/www/html/postfixadmin/model/PFASmarty.php on line 137
Deprecated: htmlentities(): Passing null to parameter #1 ($string) of type string is deprecated in /var/www/html/postfixadmin/model/PFASmarty.php on line 137
Deprecated: htmlentities(): Passing null to parameter #1 ($string) of type string is deprecated in /var/www/html/postfixadmin/model/PFASmarty.php on line 137
Invalid token (session timeout; refresh the page and try again?)

Returned option error_reporting = E_ALL & ~E_DEPRECATED & ~E_STRICT
But these mistakes are all the time now

I decided to check Roundcube, typed in the address, entered the username and password. This line came out:
Warning: Unknown: POST Content-Length of 138 bytes exceeds the limit of 50 bytes in Unknown on line 0

but the Roundcube window does not disappear after entering the login and password like the Postfixadmin window
I don't know if Postfixadmin is outdated or something with php
But Rouncube is definitely the newest version and still one line with an error
Warning: Unknown: POST Content-Length of 138 bytes exceeds the limit of 50 bytes in Unknown on line 0

I turned cookies back on...
After the display_errors option in the morning

Please tell me how to view the version of Postfixadmin ???????
I was looking all day yesterday for how to do this and couldn’t find it anywhere

Thank you for your attention

**Comment by sovkutsan** [comment]
how to view the version of Postfixadmin ?

**Comment by DavidGoodwin** [comment]
> how to view the version of Postfixadmin ?

I believe it normally says in the footer on each page; failing that, look and see what package version you have installed (e.g. `dpkg -l | grep postfixadmin`) ?

**Comment by DavidGoodwin** [comment]
> Hello Today I changed some things, but the understanding has not yet appeared. in file etc/php/8.1/fpm/php.ini
> 
> I did: display_errors = On


Try instead this in your php.ini : 

```
error_log = /var/log/php.log
display_errors = Off
```

then as root or someone : 

`touch /var/log/php.log && chmod 777 /var/log/php.log`


then restart PHP/Apache.

Most of your error messages are because the deprecation notice has been shown from PFASmarty + htmlentities.

**Comment by DavidGoodwin** [comment]
(You will need to restart the php fpm process or Apache after each change to the php.ini file)

**Comment by sovkutsan** [comment]
Thank you very much for your help.
I turned off Apache and removed it from Ubuntu startup. Everything now works only on Nginx on port 80.

Special thanks to you for your help in determining the version of Postfixadmin.

Now I'm at home, I just turned off the server remotely. It's 1:04 am here.
I'll be at work at 8 o'clock and will try everything you said.

Judging by the errors, the outdated version of Postfixadmin for php8.1

But it seems to me that even if I install a newer Postfixadmin (3.3.13) for php8.1
Or vice versa, I’ll install php7.4 if it suddenly turns out that I have an old Postfixadmin
All the same, the CSRF verification error will remain.

I think I need to install something to configure CSRF protection. Otherwise, the https protocol will perceive any attempt I make to log in as an attack, and the CSRF check will not work.

**Comment by sovkutsan** [comment]
Hello, the problem is fixed

I created a log file as you wrote:
/var/log/php.log
chmod 777 /var/log/php.log
Disabled displaying errors on the screen:
display_errors = Off
The bugs are there now
It’s not scary at all, I read about it, for version php8.1 it’s normal and doesn’t affect anything.
(PHP Deprecated: htmlentities():)

To solve my problem everything turned out to be simple:
/etc/php/8.1/fpm/php.ini
I set a high value to two options:
post_max_size = 1000M
upload_max_filesize = 1000M

I hope this helps someone

Many thanks to Mr. DavidGoodwin
Without you, I would not be able to solve this incomprehensible problem.

Now I have another one))
